File: c/algorithms/sorts/src/merge_sort.c
#include <sort_common.h>

void merge_sort_helper(int a[], int lb, int ub);
void merge(int a[], int b[], int la, int lb);

void merge_sort (int a[], int n)
{
	merge_sort_helper(a, 0, n-1);
}


void merge_sort_helper (int *a, int lb, int ub)
{
	int mid;

	if (lb == ub)
		return;

	mid = (lb + ub)/2;

	merge_sort_helper(a, lb, mid);
	merge_sort_helper(a, mid+1, ub);

	merge((a+lb), (a+mid+1), (mid-lb+1), (ub-mid));
}


void merge (int *a, int *b, int la, int lb)
{
	int i, j, k;
	int temp_array[la+lb];

	i = j = k = 0;
	while ((i < la) && (j < lb)) {
		if (a[i] < b[j])
			temp_array[k++] = a[i++];
		else
			temp_array[k++] = b[j++];
	}

	while (i < la)
		temp_array[k++] = a[i++];

	while (j < lb)
		temp_array[k++] = b[j++];

	/** Copy temporary array back to a[] and b[] in order */
	for (k=0; k<la; k++)
		a[k] = temp_array[k];

	for (j=0; j<lb; j++)
		b[j] = temp_array[k++];
}

File: python/algorithms/sorts/merge_sort.py
#!/usr/bin/python                                                               

''' 
Implement merge sort
'''

from merge import merge


def merge_sort(a):
	merge_sort_r(a, 0, len(a)-1)
	return a


# Copy all of list 'b' into 'a' starting at a['start_idx']
def copy_list(a, start_idx, b):
	i = start_idx
	for x in b:
		a[i] = x
		i += 1



# Recursive merge-sort
def merge_sort_r(a, lb, ub):
	if lb == ub:
		return

	mid = (lb + ub)/2

	merge_sort_r(a, lb, mid)
	merge_sort_r(a, mid+1, ub)

	tmp = merge(a, lb, mid+1, a, mid+1, ub+1)

	# Copy back merged sub-array into the original list, a[lb:ub]
	copy_list(a, lb, tmp)


if __name__ == '__main__':
	assert (merge_sort(
		[5, 1, 7, 3, 4, 2, 8]) == [1, 2, 3, 4, 5, 7, 8])

	assert (merge_sort(
		[11, 9, 7, 5, 3, 2, 1]) == [1, 2, 3, 5, 7, 9, 11])

	assert (merge_sort(
		[1, 2, 3, 4, 5, 6, 7]) == [1, 2, 3, 4, 5, 6, 7])

	assert (merge_sort(
		[1, 3, 5, 7, 2, 4, 6]) == [1, 2, 3, 4, 5, 6, 7])

